internal/service/tesla.go

```diff
@@ -0,0 +1,129 @@
+package service
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/DIMO-Network/shared/pkg/cipher"
+	"github.com/DIMO-Network/shared/pkg/db"
+	"github.com/DIMO-Network/tesla-oracle/internal/config"
+	dbmodels "github.com/DIMO-Network/tesla-oracle/models"
+	"github.com/rs/zerolog"
+	"regexp"
+	"strconv"
+)
+
+type TeslaService struct {
+	settings *config.Settings
+	logger   *zerolog.Logger
+	Cipher   cipher.Cipher
+	pdb      *db.Store
+}
+
+func NewTeslaService(settings *config.Settings, logger *zerolog.Logger, cipher cipher.Cipher, pdb *db.Store) *TeslaService {
+	return &TeslaService{
+		settings: settings,
+		logger:   logger,
+		Cipher:   cipher,
+		pdb:      pdb,
+	}
+}
+
+type VirtualKeyStatus int
+
+const (
+	Incapable VirtualKeyStatus = iota
+	Paired
+	Unpaired
+)
+
+// GetVehicleByVIN retrieves a vehicle by its VIN.
+func (ts *TeslaService) GetVehicleByVIN(ctx context.Context, logger *zerolog.Logger, pdb *db.Store, vin string) (*dbmodels.SyntheticDevice, error) {
+	tx, err := pdb.DBS().Writer.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelReadCommitted})
+	if err != nil {
+		logger.Error().Err(err).Msgf("Failed to begin transaction for vehicle %s", vin)
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			if rbErr := tx.Rollback(); rbErr != nil {
+				logger.Error().Err(rbErr).Msgf("GetVehicleByVIN: Failed to rollback transaction for vehicle %s", vin)
+			}
+		} else {
+			if cmErr := tx.Commit(); cmErr != nil {
+				logger.Error().Err(cmErr).Msgf("GetVehicleByVIN: Failed to commit transaction for vehicle %s", vin)
+			}
+		}
+	}()
+
+	sd, err := dbmodels.SyntheticDevices(
+		dbmodels.SyntheticDeviceWhere.Vin.EQ(vin)).One(ctx, tx)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ErrVehicleNotFound
+		}
+		logger.Error().Err(err).Msgf("Failed to check if vehicle %s has been processed", vin)
+		return nil, err
```

Comment:
General "rule"/ "guideline" I like to follow is to either log or return an error but not both. Since if we are returning an error it's nice to let the caller determine how they want to handle it. This helps 
1. Avoid but the function and caller logging the same error twice.
2. The function logging an error when the caller is okay with the function failing creating false positives in the logs.